src/frontends/mutate.cpp

```diff
@@ -6,11 +6,96 @@ LICENSE file in the root directory of this source tree.
 */
 
 #include <loop_tool/mutate.h>
+#include <loop_tool/measure.hpp>
+#include <loop_tool/backend.h>
 
 #include <algorithm>
+#include <string>
 
 namespace loop_tool {
 
+std::vector<std::string> get_available_actions(const LoopTree& lt, LoopTree::TreeRef ref){
+  std::vector<std::string> avail_actions;
+
+  LoopTree::TreeRef p_ref = previous_ref(lt, ref);
+  LoopTree::TreeRef n_ref = next_ref(lt, ref);
+
+  if (p_ref == ref){
+    p_ref = -1;
+  }
+  if (n_ref == ref){
+    n_ref = -1;
+  }
+
+  // General operations
+  if (p_ref != -1){
+    avail_actions.push_back("up");
+  }
+  if (n_ref != -1){
+    avail_actions.push_back("down");
+  }
+
+  // Loop operations
+  if (lt.kind(ref) == LoopTree::LOOP){
+    // split
+    int max_vec_size = 1;
+
+    // the most inner-loop
+    if (all_of(lt.children(ref).begin(), lt.children(ref).end(), 
+          [&] (int i) {return lt.kind(i) == LoopTree::NODE;})
+        ){
+      max_vec_size = isa_traits<DABUN_ISA>().vector_size;
+    }
+    auto loop_iter = lt.loop(ref).size;
+    while (loop_iter >= max_vec_size){
+      avail_actions.push_back("split_" + std::to_string(loop_iter)); 
+      loop_iter /= 2; 
+    }
+
+    // merge
+    if(p_ref != -1 && lt.kind(p_ref) == LoopTree::LOOP && lt.loop(ref).var == lt.loop(p_ref).var){
+      avail_actions.push_back("merge");
+    }
+
+    // swap loops
+    if (p_ref != -1 && lt.kind(p_ref) == LoopTree::LOOP){
+      avail_actions.push_back("swap_up");
+    } 
+    if (n_ref != -1 && lt.kind(n_ref) == LoopTree::LOOP){
+      avail_actions.push_back("swap_down");
+    } 
+
+    // Loop annotations
+    avail_actions.push_back("vectorize");
+    avail_actions.push_back("unroll");
+  }
+  if (lt.kind(ref) == LoopTree::NODE){
+    
+    // swap nodes
+    if (p_ref != -1 && lt.kind(p_ref) == LoopTree::NODE){
```

Comment:
We need to check if this is topologically valid, or it won't swap

Reply:
How do we know if topology is valid?